### PR TITLE
Issue 71: Implement RFC 0004-isolate_integrator_modifications_via_porting_interfaces

### PR DESCRIPTION
**Pull request by ericeberry**

### CODE REVIEW

File: reference/README.md [hunk 1]
```diff
@@ -31,19 +31,22 @@ to the TA.
 Client implementation library is required to be ported to both the host (REE) environment for use
 by client applications, as well as the TEE environment for use by TA clients.
 
-Only the files in the 'internal' directory need to be modified.
+Only the files in the 'internal' directory need to be modified. All other code in the src directory
+is platform independent and should not be modified.
```
Comment (mhabrat) [inline reference/README.md:35]:
```suggestion
is platform independent and should *NOT* be modified.
```

File: reference/README.md [hunk 2]
```diff
@@ -31,19 +31,22 @@ to the TA.
...
+ta_close_session, ta_invoke_command, ta_alloc_shared_memory, and ta_free_shared_memory which are
+defined in ta_client.h. Example implementations are in ta_client.c. ta_client.h also defines the
+macros, controlled by the compile time USE_SHARED_MEMORY flag, that determine whether shared memory
+or standard memory is used by the client-side library.
 
-Vendors must implement code identified by ```TODO SoC Vendor```.
+Vendors *MUST* implement code identified by ```TODO SoC Vendor``` in files in the src/porting
+directory.
+
+Vendors can modify can in the src/internal if needed.
```
Comment (mhabrat) [inline reference/README.md:49]:
```suggestion
Vendors may modify can in the src/internal if needed.
```

File: reference/README.md [hunk 3]
```diff
@@ -31,19 +31,22 @@ to the TA.
 Client implementation library is required to be ported to both the host (REE) environment for use
 by client applications, as well as the TEE environment for use by TA clients.
 
-Only the files in the 'internal' directory need to be modified.
+Only the files in the 'internal' directory need to be modified. All other code in the src directory
+is platform independent and should not be modified.
 
 clientimpl code performs the marshaling of calls from API calls into the TA. Some TEEs require
 communication through shared memory, others may be able to use standard memory between the processor
 and the TEE. To use shared memory, define the compile time flag USE_SHARED_MEMORY. Vendors must
```
Comment (mhabrat) [inline reference/README.md:39]:
```suggestion
and the TEE. To use shared memory, define the compile time flag USE_SHARED_MEMORY. Vendors *MUST*
```

File: reference/README.md [hunk 4]
```diff
@@ -54,7 +57,15 @@ Only the files in the include/internal, include/porting, src/internal, and src/p
 need to be modified. All other code in the src and include directories is platform independent and
 should not be modified.
```
Comment (mhabrat) [inline reference/README.md:58]:
```suggestion
should *NOT* be modified.
```

File: reference/README.md [hunk 5]
```diff
@@ -54,7 +57,15 @@ Only the files in the include/internal, include/porting, src/internal, and src/p
...
 
-Vendors must implement code identified by ```TODO SoC Vendor```.
+Vendors must implement code to call the TA-side functions: ta_open_session_handler,
```
Comment (mhabrat) [inline reference/README.md:60]:
```suggestion
Vendors *MUST* implement code to call the TA-side functions: ta_open_session_handler,
```

File: reference/README.md [hunk 6]
```diff
@@ -54,7 +57,15 @@ Only the files in the include/internal, include/porting, src/internal, and src/p
...
+ta_close_session_handler, and ta_invoke_command_handler which defined in ta.h and implemented in
+ta.c.
+
+Vendors *MUST* implement code identified by ```TODO SoC Vendor``` in files in the include/porting and
+src/porting directories.
+
+Vendors can modify in the include/internal and src/internal if needed. For instance to replace
```
Comment (mhabrat) [inline reference/README.md:67]:
```suggestion
Vendors may modify in the include/internal and src/internal if needed. For instance to replace
```

File: reference/README.md [hunk 7]
```diff
@@ -18,8 +18,8 @@ SecAPI.
 This folder should be copied over from the reference implementation to the SoC vendor
 implementation as-is, without any changes. Comcast will update the reference implementation and
 unit tests over time and will expect SoC vendors to keep this folder up to date. Some unit tests,
-like sa_key_import_soc.cpp, may need to be modified by the vendor, especially if the vendor library
-does not support a particular feature. The vendor must declare to Comcast which unit tests have been
+like sa_key_import_soc.cpp, may be modified by the vendor, especially if the vendor library does
+not support a particular feature. The vendor must declare to Comcast which unit tests have been
```
Comment (mhabrat) [inline reference/README.md:22]:
```suggestion
not support a particular feature. The vendor *MUST* declare to Comcast which unit tests have been
```

File: reference/src/taimpl/src/porting/transport.c [hunk 8]
```diff
@@ -38,13 +39,13 @@ sa_status transport_authenticate_caller(sa_uuid* uuid) {
 
     // SecApi TA obtains the caller UUID from the underlying transport mechanism which is
     // platform dependent. The caller must be authenticated to prevent spoofing.
-    // TODO Soc Vendor: add code here to retrieve the authenticated caller UUID.
-
+    // TODO Soc Vendor: replace code here to retrieve the authenticated caller UUID. If the caller is the REE, return
```
Comment (mhabrat) [inline reference/src/taimpl/src/porting/transport.c:42]: Why not create a function that retrieves the authenticated caller UUID and place the new function in the porting layer for the SOC manager to define? 

In that spirit could all code required to be modified by SOC vendors be abstracted into a set of functions that are defined in the porting directory, isolating all SOC vendor changes to a specific layer/directory (porting) and eliminating the need for them to modify code in the internal directory?

**Review by mhabrat — CHANGES_REQUESTED** [review]

File: reference/src/taimpl/src/porting/transport.c (hunk 8 above)
Comment (mhabrat) [reply reference/src/taimpl/src/porting/transport.c:42]: I recommend adding these additional details to the README.md in the reference directory. The reason for my comment is to try and simplify the porting interface that the SOC vendor has to maintain to a single directory. As additional SOC vendors adopt SecAPI, we should ensure the process is as clear and painless as possible to reduce porting time and ensure a consistent and secure implementation.

File: reference/README.md [hunk 9]
```diff
@@ -52,9 +55,17 @@ servicing client requests.
...
 Only the files in the include/internal, include/porting, src/internal, and src/porting directories
 need to be modified. All other code in the src and include directories is platform independent and
-should not be modified.
+should *NOT* be modified.
+
+Vendors *MUST* implement code to call the TA-side functions: ta_open_session_handler,
+ta_close_session_handler, and ta_invoke_command_handler which defined in ta.h and implemented in
+ta.c.
+
+Vendors *MUST* implement code identified by ```TODO SoC Vendor``` in files in the include/porting and
```
Comment (mhabrat) [inline reference/README.md:64]: Recommend checking and updating the code as I saw at least one instance where the tag was TODO Soc Vendor.

**Review by mhabrat — CHANGES_REQUESTED** [review]

File: reference/src/taimpl/src/porting/transport.c (hunk 8 above)
Comment (ericeberry) [reply reference/src/taimpl/src/porting/transport.c]: Done

**Review by mhabrat — DISMISSED** [review]

**Review by mhabrat — APPROVED** [review]

**Pull request merged**